### ISSUE TITLE
Uninstalling tigera-operator helm chart fails with "tigera-operator-uninstall" job.

### ISSUE THREAD

**Issue by MoneshKumar** [projectcalico/calico#8436]
Installed tigera-operator helm chart from below link, and everything went fine during installation.
https://github.com/projectcalico/calico/tree/master
However, getting error while uninstalling the helm chart.

**Note :** Copied only directories '[charts/tigera-operator/](https://github.com/projectcalico/calico/tree/master/charts/tigera-operator)' and '[libcalico-go/](https://github.com/projectcalico/calico/tree/master/libcalico-go)' from the above repo into my environment.

## Expected Behavior
All the calico pods and tigera-operator pod should be deleted.

## Current Behavior
- **tigera-operator-uninstall** job is getting failed. 
	Below is the logs of helm chart uninstallation.
       <img width="639" alt="uninstall_error_4" src="https://github.com/projectcalico/calico/assets/12606141/0c57e630-e389-41dc-afc5-6d06518361cc">

	Below is the logs of **tigera-operator-uninstall** pod.
> kubectl logs tigera-operator-uninstall-n2qvt -n infra-services
> 2024/01/23 15:09:03 [INFO] Version: c68c5cc2f2f5
> 2024/01/23 15:09:03 [INFO] Go Version: go1.21.6 X:boringcrypto
> 2024/01/23 15:09:03 [INFO] Go OS/Arch: linux/amd64
> 2024/01/23 15:09:04 [INFO] Waiting for Installation to be fully deleted
> 2024/01/23 15:13:55 [INFO] Waiting for Installation to be fully deleted
> 2024/01/23 15:14:00 [INFO] Waiting for Installation to be fully deleted
> 2024/01/23 15:14:05 [INFO] preDelete hook exiting
> 2024/01/23 15:14:05 [ERROR] Timeout waiting for pre-delete hookFailed to complete pre-delete hook

- **calico-typha** pods under **calico-system** namespace were left.
> kubectl get po -A | grep 'calico\|tigera'
> calico-system    calico-typha-59c6847bc7-4rj26                             1/1     Running                 0                36m
> calico-system    calico-typha-59c6847bc7-m6tgb                             1/1     Running                 0                36m
> calico-system    calico-typha-59c6847bc7-rv57f                             1/1     Running                 0                36m
> infra-services   tigera-operator-5cc9b4b697-thcgj                          1/1     Running                 0                37m
> infra-services   tigera-operator-uninstall-fpdpd                           0/1     Completed               0                46s
> infra-services   tigera-operator-uninstall-n2qvt                           0/1     Error                   0                5m53s

Note: Earlier other pods(apiserver) also were left during uninstallation. But with [last week's updated repo](https://github.com/projectcalico/calico/commit/7f5db2347fc5f65ff74c67f7b588c8d9509f3dbe#diff-b4c86961bf65b882d114badaf57dd38c993a28d700480db86bbff0b4ff299b84), can see only the calico-typha pods were not getting removed now.
	
- ServiceAccounts and crds were also left.

## Your Environment
* Calico version: v3.27.0
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes/EKS 1.26
* Operating System and version: CentOS7

**Comment by caseydavenport** [comment]
I think we'll need to wait and see how this works with an official build of the helm chart before making too much judgement. I'll also give this a spin in my environment and see if I can reproduce it.

**Comment by caseydavenport** [comment]
I managed to reproduce this on my cluster - the problem appears to be that the `calico-apiserver` Namespace is never cleaned up, and the Installation controller is waiting for it to be removed before it continues.

We may need to add a small bit of explicit code to delete this resource if for some reason the Kubernetes GC isn't handling it as expected.

**Comment by caseydavenport** [comment]
There seems to be another issue with the Kubernetes garbage collection logic that leads to this. Not sure where yet exactly, but if I trigger a trivial change on one of these resources (namespace, serviceaccount, etc) like adding an annotation, the garbage collector seems to notice the change right away and deletes the resource. Without that trigger, it seems to get lost.

We can probably add some explicit code to Calico to handle this, but sounds like it might be a bug in upstream Kubernetes.

**Comment by stiyyagura0901** [comment]
Facing a similar issue with v3.27.0. Logs from `tigera-operator/tigera-operator-uninstall-6xjbb:cleanup-job`

```
2024/01/25 06:03:02 [INFO] Version: v1.32.3                                                                                                                                                │
│ 2024/01/25 06:03:02 [INFO] Go Version: go1.21.5 X:boringcrypto                                                                                                                             │
│ 2024/01/25 06:03:02 [INFO] Go OS/Arch: linux/amd64                                                                                                                                         │
│ 2024/01/25 06:03:03 [INFO] preDelete hook exiting                                                                                                                                          │
│ 2024/01/25 06:03:03 [ERROR] installations.operator.tigera.io "default" is forbidden: User "system:serviceaccount:tigera-operator:tigera-operator" cannot delete resource "installations" i │
│ n API group "operator.tigera.io" at the cluster scopeFailed to complete pre-delete hook                                                                                                    │
│ Stream closed EOF for tigera-operator/tigera-operator-uninstall-6xjbb (cleanup-job)
```

however, [this](https://github.com/projectcalico/calico/blob/master/charts/tigera-operator/templates/tigera-operator/02-role-tigera-operator.yaml#L126) seems to be added recently but not released yet.
It is not present in [v3.28.0-0.dev](https://github.com/projectcalico/calico/blob/v3.28.0-0.dev/charts/tigera-operator/templates/tigera-operator/02-role-tigera-operator.yaml) as well.

**Comment by caseydavenport** [comment]
@stiyyagura0901 that fix will be included in v3.27.1 (hopefully soon) as well as v3.28.0 when it is released in a few months.

**Comment by MoneshKumar** [comment]
@caseydavenport, thank you for your analysis, I would like to know if this cleanup issue will be considered in next chart release, and if yes, when can be the release..

**Comment by caseydavenport** [comment]
I've been out of the office for a few weeks, will catch up on this one and see what I can do. It's not obvious what the right fix is at this point.

**Comment by caseydavenport** [comment]
This should be fixed in the upcoming Calico v3.28 release. This PR fixed the issue that was causing namespaces / serviceaccounts to not be properly deleted: https://github.com/projectcalico/calico/pull/8586
